### ISSUE TITLE
The underscore "_" is a reserved token in Java 9. The jOOQ code generator should avoid generating it

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#4703]
When a column name consists of a single whitespace or other special character:

``` sql
SELECT 1 AS " ", 2 AS "-"
```

jOOQ's code generator will generate an underscore as the column name. With Java 9, the single underscore becomes a reserved token and must thus no longer be used.

**Comment by dfraser** [comment]
I have an enum with _ as a value.  Bad news in Java 9 :)

**Comment by lukaseder** [comment]
Interesting, what's the use-case? :)

**Comment by dfraser** [comment]
It's something silly, like UI menu items being represented by a database enum. Indubitably a bad idea, but the codegen should avoid making invalid identifiers... 🙃  Even if it skipped it with a warning.

**Comment by lukaseder** [comment]
Oh I see, your enum is generated. Of course, this issue should fix this problem. The workaround is to implement a generator strategy:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-generatorstrategy

**Comment by palindrom615** [comment]
I am new with jOOQ and I wonder there is any news.

Actually, I too have trouble on this now. When jooq generate code of `information_schema` schema, which is mysql's meta table, there exists enum value of empty string. `UserSslType` of `User` table is. and `ColumnsColumnKey` of 'columns` table is another.

There's only two types, so I can simply change the property with some other name. but I wonder how on earth jooq with java >= 9 & mysql user avoid this. Or, they just ignore the issue?

**Comment by lukaseder** [comment]
@palindrom615 You can easily work around this problem either by implementing a generator strategy (see previous comments and the manual), or in your case, by excluding the schemas / tables that produce this problem. I'm assuming you don't really need to generate the `information_schema`?

**Comment by palindrom615** [comment]
Thanks for quick reply, @lukaseder ! I used `include` option to generate db only what I need and it works great.

**Comment by cghislai** [comment]
@lukaseder Unfortunately the naming strategy does not seem to be called for enum constants.
In `org.jooq.codegen.GenerationUtil`, you have the following snippet:

```
    public static String convertToIdentifier(String literal, Language language) {
        if (language == JAVA && JAVA_KEYWORDS.contains(literal))
            return literal + "_";
        if (language == SCALA && SCALA_KEYWORDS.contains(literal))
            return "`" + literal + "`";

        StringBuilder sb = new StringBuilder();

        if ("".equals(literal))
            if (language == SCALA)
                return "`_`";
            else
                return "_";
```

**Comment by lukaseder** [comment]
Thanks for your hint. There is an issue for that as well, but I cannot seem to find it :)

You can always override `JavaGenerator.generateEnum()` methods, or patch it. Sorry, I don't have a workaround for these problems right now.

**Comment by cghislai** [comment]
For those interested, here is an generator implementation working around this issue for Java code generation of a mariadb schema. Note that you may need to adapt it if you enum literals contains reserved keywords for instance.

<details>
<summary>
Java9SafeEnumIdentifier
</summary>

```java
public class ValidJava9EnumConstantsGenerator extends JavaGenerator {

    @Override
    protected void generateEnum(EnumDefinition e, JavaWriter out) {
        final String className = getStrategy().getJavaClassName(e, GeneratorStrategy.Mode.ENUM);
        final List<String> interfaces = out.ref(getStrategy().getJavaClassImplements(e, GeneratorStrategy.Mode.ENUM));
        final List<String> literals = e.getLiterals();
        final List<String> identifiers = new ArrayList<String>(literals.size());

        for (int i = 0; i < literals.size(); i++) {
            String literal = literals.get(i);
            String baseIdentifier = literal.replaceAll("-", "_")
                    .toUpperCase();
            String safeIdentifier = "_" + baseIdentifier + "_VALUE";

            // [#2781] Disambiguate collisions with the leading package name
            if (safeIdentifier.equals(getStrategy().getJavaPackageName(e).replaceAll("\\..*", "")))
                safeIdentifier += "_";

            identifiers.add(safeIdentifier);
        }

        printPackage(out, e);
        generateEnumClassJavadoc(e, out);
        printClassAnnotations(out, e.getSchema());


        boolean enumHasNoSchema = e.isSynthetic() || !(e.getDatabase() instanceof PostgresDatabase);

        interfaces.add(out.ref(EnumType.class));
        out.println("public enum %s[[before= implements ][%s]] {", className, interfaces);

        for (int i = 0; i < literals.size(); i++) {
            out.println();
            out.tab(1).println("%s(\"%s\")%s", identifiers.get(i), literals.get(i), (i == literals.size() - 1) ? ";" : ",");
        }

        out.println();
        out.tab(1).println("private final %s literal;", String.class);
        out.println();
        out.tab(1).println("private %s(%s literal) {", className, String.class);
        out.tab(2).println("this.literal = literal;");
        out.tab(1).println("}");

        out.tab(1).overrideInherit();
        out.tab(1).println("public %s getCatalog() {", Catalog.class);

        if (enumHasNoSchema)
            out.tab(2).println("return null;");
        else
            out.tab(2).println("return getSchema() == null ? null : getSchema().getCatalog();");

        out.tab(1).println("}");

        // [#2135] Only the PostgreSQL database supports schema-scoped enum types
        out.tab(1).overrideInherit();
        out.tab(1).println("public %s getSchema() {", Schema.class);
        out.tab(2).println("return null;");
        out.tab(1).println("}");

        out.tab(1).overrideInherit();
        out.tab(1).println("public %s getName() {", String.class);
        out.tab(2).println("return %s;", e.isSynthetic() ? "null" : "\"" + e.getName().replace("\"", "\\\"") + "\"");
        out.tab(1).println("}");

        out.tab(1).overrideInherit();
        out.tab(1).println("public %s getLiteral() {", String.class);
        out.tab(2).println("return literal;");
        out.tab(1).println("}");

        generateEnumClassFooter(e, out);
        out.println("}");
    }
}

```
</details>

**Comment by lukaseder** [comment]
At some point, we need a much better collision strategy than what we have now. For now, if there's an underscore only identifiers, we'll just add one more underscore. This prevents collisions in schemas like this:

```sql
create table _ (_ int);
create table __ (__ int);
```

**Comment by lukaseder** [comment]
This actually doesn't work yet, there are collisions between generated tables / columns, and parameter names:

```java
create table _ (_ int);
create table __ (__ int);
```

I won't fix this for now. If someone has this in their schema, they're asking for trouble.

**Comment by lukaseder** [comment]
There's a slight risk that this breaks someone's code if they already have all-underscore identifiers. This is why I won't backport the fix. Fixed for jOOQ 3.14.0
